xc7/utils/prjxray_form_channels.py

```diff
@@ -929,6 +986,160 @@ def create_constant_tracks(conn):
+            pip_pkey_map = c.execute("SELECT name, pkey FROM pip_in_tile WHERE tile_type_pkey = (?)", (tile_type_pkey,)).fetchall()
+            pip_pkey_map = {m[0]: m[1] for m in pip_pkey_map}
+
+            # Insert tiles into the VPR grid
+            for vpr_tile_type_pkey, loc_ofs in tile_map.fwd_map[tile_type_pkey]:
+
+                # Compute location in the VPR grid
+                vpr_loc = (
+                    base_vpr_loc[0] + loc_ofs[0],
+                    base_vpr_loc[1] + loc_ofs[1],
+                )
+
+                # Get tile type as string
+                vpr_tile_type = c.execute("SELECT name FROM tile_type WHERE pkey = (?)", (vpr_tile_type_pkey, )).fetchone()[0]
+
+                # Generate new tile name
+                # FIXME: This will be the SLICE name. However it will not match
+                # FIXME: the Vivado slice name, coordinates will differ.
+                vpr_tile_name = "%s_X%dY%d" % (vpr_tile_type, vpr_loc[0], vpr_loc[1])
+
+                # Insert the tile into grid
+                vpr_tile_pkey = insert_vpr_tile(conn, vpr_tile_name, vpr_loc, vpr_tile_type_pkey, tile[0])
+
+                # Get site instance basing on the new tile type and site
+                # location offset. Here we assume that new tile type is
+                # equal to its former site type.
+                site_name, site_pkey = c.execute("SELECT name, pkey FROM site WHERE x_coord = (?) AND y_coord = (?) AND site_type_pkey = (SELECT pkey FROM site_type WHERE name = (?))", (loc_ofs[0], loc_ofs[1], vpr_tile_type)).fetchone()
+                assert site_pkey is not None
+
+                # Add entry to the tile_former_site table
+                c.execute("INSERT INTO tile_former_site(vpr_tile_pkey, site_pkey) VALUES (?, ?)", (vpr_tile_pkey, site_pkey))
+
+                # Insert correspondencies to the pip_in_tile_instance table.
+                for pip_name in pip_map[tile_type].fwd_map[site_name]:
+                    #print(pip_name)
+
+                    # FIXME: TODO:
+                    # Skip ppips in a very very hackish way for now.
+                    # All ppips have '->>' in their names
+                    if "->>" in pip_name:
+                        continue
+
+                    # Get pip_in_tile pkey
+                    pip_pkey = pip_pkey_map[pip_name]
+                    assert pip_pkey is not None
+
+                    # Insert
+                    c.execute("INSERT INTO pip_in_tile_instance(vpr_tile_pkey, pip_in_tile_pkey) VALUES (?, ?)", (vpr_tile_pkey, pip_pkey))
+
+        # The tile is not being split
+        else:
+
+            # Insert the tile into grid
+            insert_vpr_tile(conn, tile[1], vpr_locs[0], tile_type_pkey, tile[0])
+
+            # If one physical location corresponds to more than one VPR location
+            # then fill the space with artificial EMPTY tiles.
+            for i in range(1, len(vpr_locs)):
```

Comment:
In what case does this ever occur?  I think ``assert len(vpr_locs) == 1`` makes more sense here.

Reply:
While doing the CLB split the entire grid column is being divided into two. There are tiles which are not CLB in such a column and those are not split. So in the end we would have "holes" in the grid. And this piece of code inserts tiles there of NULL type.

Though I'm not sure whether this is actually needed. Are such "holes" in the grid ok? I mean for channel/track formation etc?

Reply:
I understand what is happening, and I think this structure is confusing.  I think the tile splitter should generate a tile map AND a "filler" tile list.  Do not add filler tiles to the unmodified tiles, just add a "filler list".

So the logic looks like:
for live tiles:
  if map:
    output mapped tiles
  else:
    output moved tile (1 to 1)

for filler tiles:
  add_filler_tile